moshi/src/main/java/com/squareup/moshi/JsonWriter.kt

```diff
@@ -0,0 +1,548 @@
+ * object's properties by alternating calls to [name] with the property's value. Write
+ * property values with the appropriate [value] method or by nesting other objects or
+ * arrays. Finally close the object using [endObject].
+ *
+ *
+ * ## Example
+ *
+ * Suppose we'd like to encode a stream of messages such as the following:
+ *
+ * ```json
+ * [
+ *   {
+ *     "id": 912345678901,
+ *     "text": "How do I stream JSON in Java?",
+ *     "geo": null,
+ *     "user": {
+ *       "name": "json_newb",
+ *       "followers_count": 41
+ *     }
+ *   },
+ *   {
+ *     "id": 912345678902,
+ *     "text": "@json_newb just use JsonWriter!",
+ *     "geo": [
+ *       50.454722,
+ *       -104.606667
+ *     ],
+ *     "user": {
+ *       "name": "jesse",
+ *       "followers_count": 2
+ *     }
+ *   }
+ * ]
+ * ```
+ *
+ * This code encodes the above structure:
+ *
+ * ```java
+ * public void writeJsonStream(BufferedSink sink, List<Message> messages) throws IOException {
+ *   JsonWriter writer = JsonWriter.of(sink);
+ *   writer.setIndent("  ");
+ *   writeMessagesArray(writer, messages);
+ *   writer.close();
+ * }
+ *
+ * public void writeMessagesArray(JsonWriter writer, List<Message> messages) throws IOException {
+ *   writer.beginArray();
+ *   for (Message message : messages) {
+ *     writeMessage(writer, message);
+ *   }
+ *   writer.endArray();
+ * }
+ *
+ * public void writeMessage(JsonWriter writer, Message message) throws IOException {
+ *   writer.beginObject();
+ *   writer.name("id").value(message.getId());
+ *   writer.name("text").value(message.getText());
+ *   if (message.getGeo() != null) {
+ *     writer.name("geo");
+ *     writeDoublesArray(writer, message.getGeo());
+ *   } else {
+ *     writer.name("geo").nullValue();
+ *   }
+ *   writer.name("user");
+ *   writeUser(writer, message.getUser());
+ *   writer.endObject();
+ * }
+ *
+ * public void writeUser(JsonWriter writer, User user) throws IOException {
+ *   writer.beginObject();
+ *   writer.name("name").value(user.getName());
+ *   writer.name("followers_count").value(user.getFollowersCount());
+ *   writer.endObject();
+ * }
+ *
+ * public void writeDoublesArray(JsonWriter writer, List<Double> doubles) throws IOException {
+ *   writer.beginArray();
+ *   for (Double value : doubles) {
+ *     writer.value(value);
+ *   }
+ *   writer.endArray();
+ * }
+ * ```
+ *
+ *
+ * Each `JsonWriter` may be used to write a single JSON stream. Instances of this class are
+ * not thread safe. Calls that would result in a malformed JSON string will fail with an [IllegalStateException].
+ */
+public abstract class JsonWriter internal constructor() : Closeable, Flushable {
+  /**
+   * The nesting stack. Using a manual array rather than an ArrayList saves 20%. This stack will
+   * grow itself up to 256 levels of nesting including the top-level document. Deeper nesting is
+   * prone to trigger StackOverflowErrors.
+   */
+  @JvmField
+  public var stackSize: Int = 0
```

Comment:
These package-private fields should be `internal` or `protected`. Start with the latter as we'll want to make JsonWriter `sealed` eventually

Reply:
2ed413bd2fd1ff66372e2444ae78cc3e91086d50